transport/roundtrip_test.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/stretchr/testify/require"
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/testutils"
+)
+
+// all tests in this file should use these names for callers and services.
+const (
+	testCaller  = "testService-client"
+	testService = "testService"
+)
+
+// roundTripTransport provides a function that sets up and tears down an
+// Inbound, and provides an Outbound which knows how to call that Inbound.
+type roundTripTransport interface {
+	// Set up an Inbound serving Handler h, and call f with an Outbound that
+	// knows how to talk to that Inbound.
+	WithHandler(h transport.Handler, f func(transport.Outbound))
+}
+
+// handlerFunc wraps a function into a transport.Handler
+type handlerFunc func(context.Context, *transport.Request, transport.ResponseWriter) error
+
+func (f handlerFunc) Handle(ctx context.Context, r *transport.Request, w transport.ResponseWriter) error {
+	return f(ctx, r, w)
+}
+
+// httpTransport implements a roundTripTransport for HTTP.
+type httpTransport struct{ t *testing.T }
+
+func (ht httpTransport) WithHandler(h transport.Handler, f func(transport.Outbound)) {
+	i := http.NewInbound("127.0.0.1:0")
+	require.NoError(ht.t, i.Start(h), "failed to start")
+	defer i.Stop()
+
+	addr := fmt.Sprintf("http://%v/", i.Addr().String())
+	o := http.NewOutbound(addr)
+	f(o)
+}
+
+// tchannelTransport implements a roundTripTransport for TChannel.
+type tchannelTransport struct{ t *testing.T }
+
+func (tt tchannelTransport) WithHandler(h transport.Handler, f func(transport.Outbound)) {
+	serverOpts := testutils.NewOpts().SetServiceName(testService)
+	clientOpts := testutils.NewOpts().SetServiceName(testCaller)
+	testutils.WithServer(tt.t, serverOpts, func(ch *tchannel.Channel, hostPort string) {
+		i := tch.NewInbound(ch)
+		require.NoError(tt.t, i.Start(h), "failed to start")
+		defer i.Stop()
+		// ^ the server is already listening so this will just set up the
+		// handler.
+
+		client := testutils.NewClient(tt.t, clientOpts)
+		o := tch.NewOutbound(client, tch.HostPort(hostPort))
+
+		f(o)
+	})
+}
+
+func TestSimpleRoundTrip(t *testing.T) {
+	tests := []roundTripTransport{
+		httpTransport{t},
+		tchannelTransport{t},
+	}
+
+	request := &transport.Request{
+		Caller:    testCaller,
+		Service:   testService,
+		Procedure: "hello",
+		Encoding:  raw.Encoding,
+		Headers:   transport.Headers{"token": "1234"},
+		TTL:       200 * time.Millisecond, // TODO use default
+		Body:      bytes.NewReader([]byte("world")),
+	}
+	requestMatcher := transporttest.NewRequestMatcher(t, request)
+
+	// Matches the response that we send from the fake handler
+	responseMatcher := transporttest.NewResponseMatcher(t, &transport.Response{
+		Headers: transport.Headers{"status": "ok"},
+		Body:    ioutil.NopCloser(bytes.NewReader([]byte("hello, world"))),
+	})
+
+	h := handlerFunc(func(_ context.Context, r *transport.Request, w transport.ResponseWriter) error {
+		require.True(t, requestMatcher.Matches(r), "request did not match")
```

Comment:
do `assert.True`, `require` doesn't work outside of the test goroutine and the handler is running in a separate goroutine + you don't need to fail the test on error here anyway. (same below, change require to assert).

also print the request you received, and ideally print the request that you were looking for.